cmd/testscript/main.go

```diff
@@ -0,0 +1,217 @@
+	fs.Usage = func() {
+		mainUsage(os.Stderr)
+	}
+	fVerbose := fs.Bool("v", false, "run tests verbosely")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		return err
+	}
+
+	td, err := ioutil.TempDir("", "testscript")
+	if err != nil {
+		return fmt.Errorf("unable to create temp dir: %v", err)
+	}
+	defer os.RemoveAll(td)
+
+	files := fs.Args()
+	if len(files) == 0 {
+		files = []string{"-"}
+	}
+
+	for i, fileName := range files {
+		// TODO make running files concurrent by default? If we do, note we'll need to do
+		// something smarter with the runner stdout and stderr below
+		runDir := filepath.Join(td, "dir"+strconv.Itoa(i))
+		if err := os.Mkdir(runDir, 0777); err != nil {
+			return fmt.Errorf("failed to create a run directory within %v for %v: %v", td, fileName, err)
+		}
+		if err := run(runDir, fileName, *fVerbose); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var (
+	failedRun = errors.New("failed run")
+	skipRun   = errors.New("skip")
+)
+
+type runner struct {
+	verbose bool
+}
+
+func (r runner) Skip(is ...interface{}) {
+	panic(skipRun)
+}
+
+func (r runner) Fatal(is ...interface{}) {
+	r.Log(is...)
+	r.FailNow()
+}
+
+func (r runner) Parallel() {
+	// No-op for now; we are currently only running a single script in a
+	// testscript instance.
+}
+
+func (r runner) Log(is ...interface{}) {
+	fmt.Print(is...)
+}
+
+func (r runner) FailNow() {
+	panic(failedRun)
+}
+
+func (r runner) Run(n string, f func(t testscript.T)) {
+	// For now we we don't top/tail the run of a subtest. We are currently only
+	// running a single script in a testscript instance, which means that we
+	// will only have a single subtest.
+	f(r)
+}
+
+func (r runner) Verbose() bool {
+	return r.verbose
+}
+
+func run(runDir, fileName string, verbose bool) error {
+	var ar *txtar.Archive
+	var err error
+
+	testdata := filepath.Join(runDir, "testdata")
+	mods := filepath.Join(testdata, goModProxyDir)
+
+	if err := os.MkdirAll(mods, 0777); err != nil {
+		return fmt.Errorf("failed to create testdata dir: %v", err)
+	}
+
+	if fileName == "-" {
+		byts, err := ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("failed to read from stdin: %v", err)
+		}
+		ar = txtar.Parse(byts)
+	} else {
+		ar, err = txtar.ParseFile(fileName)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to txtar parse %v: %v", fileName, err)
+	}
+
+	var script, gomodProxy txtar.Archive
+	script.Comment = ar.Comment
+
+	for _, f := range ar.Files {
+		fp := filepath.Clean(filepath.FromSlash(f.Name))
+		parts := strings.Split(fp, string(os.PathSeparator))
+
+		if len(parts) > 1 && parts[0] == goModProxyDir {
+			gomodProxy.Files = append(gomodProxy.Files, f)
+		} else {
+			script.Files = append(script.Files, f)
+		}
+	}
+
+	if txtar.Write(&gomodProxy, testdata); err != nil {
+		return fmt.Errorf("failed to write .gomodproxy files: %v", err)
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(testdata, "script.txt"), txtar.Format(&script), 0666); err != nil {
```

Comment:
For stdin we could use stdin.txt

For real files, it's a bit tricky, yes. Because you might have passed two scripts, each with the same base name, but in different directories. So using the base name of the files passed in isn't perfect either. I'm tempted to go for a munged version of the name of the input files, and use that as the directory name (instead of `dirN`). Will push that up for thoughts.

Reply:
Actually, I've decided against the munging. It's just too messy. Instead the last line gives the file name as provided (or `<stdin>`) and tells you where it was run. I've also removed the "dir" part of `dirN`, so that 0-indexed directories within the temporary working directory are created for each file provided.

The `-work` flag (not advertised for now) prints that temporary working directory and does not remove it when the command exits (to allow you to poke around).

If the number-based naming of subdirectories proves to be annoying we can fix that down the line I guess.